applications/RomApplication/custom_utilities/rom_residuals_utility.h

```diff
@@ -0,0 +1,185 @@
+
+        mNodalDofs = mNodalVariablesNames.size();
+        mRomDofs = ThisParameters["number_of_rom_dofs"].GetInt();
+
+        // Setting up mapping: VARIABLE_KEY --> CORRECT_ROW_IN_BASIS
+        for(int k=0; k<mNodalDofs; k++){
+            if(KratosComponents<Variable<double>>::Has(mNodalVariablesNames[k]))
+            {
+                const auto& var = KratosComponents<Variable<double>>::Get(mNodalVariablesNames[k]);
+                MapPhi[var.Key()] = k;
+            }
+            else if(KratosComponents<ModelPart::VariableComponentType>::Has(mNodalVariablesNames[k]))
+            {
+                const auto& var = KratosComponents<ModelPart::VariableComponentType>::Get(mNodalVariablesNames[k]);
+                MapPhi[var.Key()] = k;
+            }
+            else
+                KRATOS_ERROR << "variable \""<< mNodalVariablesNames[k] << "\" not valid" << std::endl;
+
+        }
+    }
+
+        ~RomResidualsUtility()= default;
+
+
+        void GetPhiElemental(
+            Matrix &PhiElemental,
+            const Element::DofsVectorType &dofs,
+            const Element::GeometryType &geom)
+        {
+            const auto *current_rom_nodal_basis = &(geom[0].GetValue(ROM_BASIS));
+            int counter = 0;
+            for(unsigned int k = 0; k < dofs.size(); ++k){
+                auto variable_key = dofs[k]->GetVariable().Key();
+                if(k==0)
+                    current_rom_nodal_basis = &(geom[counter].GetValue(ROM_BASIS));
+                else if(dofs[k]->Id() != dofs[k-1]->Id()){
+                    counter++;
+                    current_rom_nodal_basis = &(geom[counter].GetValue(ROM_BASIS));
+                }
+                if (dofs[k]->IsFixed())
+                    row(PhiElemental, k) = ZeroVector(PhiElemental.size2());
+                else
+                    row(PhiElemental, k) = row(*current_rom_nodal_basis, MapPhi[variable_key]);
+            }
+        }
+
+
+        Matrix Calculate()
+        {
+            // Getting the number of elements and conditions from the model
+            const int nelements = static_cast<int>(mpModelPart.Elements().size());
+            const int nconditions = static_cast<int>(mpModelPart.Conditions().size());
+
+            auto& CurrentProcessInfo = mpModelPart.GetProcessInfo();
+            auto el_begin = mpModelPart.ElementsBegin();
+            auto cond_begin = mpModelPart.ConditionsBegin();
+
+            //contributions to the system
+            Matrix LHS_Contribution = ZeroMatrix(0, 0);
+            Vector RHS_Contribution = ZeroVector(0);
+
+            //vector containing the localization in the system of the different terms
+            Element::EquationIdVectorType EquationId;
+            Matrix MatrixResiduals( (nelements + nconditions), mRomDofs); // Matrix of reduced residuals.
+            Matrix PhiElemental;
+            for (int k = 0; k < nelements; k++){
+                auto it_el = el_begin + k;
```

Comment:
Raul I highly recommend that you create a templated function for this, where you can pass Element or Condition (and maybe Constraints in the future) this would reduce code duplication a lot!